oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -0,0 +1,334 @@
+                GrpcServerError::RequestProcessingError
+            })?;
+
+        Ok(response_reader)
+    }
+
+    /// Processes a gRPC response from a channel represented by `response_reader` and returns an
+    /// HTTP response body.
+    fn process_response(&self, response_reader: Handle) -> Result<Vec<u8>, GrpcServerError> {
+        let read_status = self
+            .runtime
+            .wait_on_channels(&[Some(response_reader)])
+            .map_err(|error| {
+                error!("Couldn't wait on the temporary gRPC channel: {:?}", error);
+                GrpcServerError::ResponseProcessingError
+            })?;
+
+        if read_status[0] == ChannelReadStatus::ReadReady {
+            self.runtime
+                .channel_read(response_reader)
+                .map_err(|error| {
+                    error!("Couldn't read temporary gRPC channel: {:?}", error);
+                    GrpcServerError::ResponseProcessingError
+                })
+                .map(|message| {
+                    // Return an empty HTTP body if the `message` is None.
+                    message.map_or(vec![], |m| m.data)
+                })
+        } else {
+            error!("Couldn't read channel: {:?}", read_status[0]);
+            Err(GrpcServerError::ResponseProcessingError)
+        }
+    }
+}
+
+/// Oak Node implementation for the gRPC server.
+impl super::Node for GrpcServerNode {
+    fn start(&mut self) -> Result<(), OakStatus> {
+        let server = self.clone();
+        let thread_handle = thread::Builder::new()
+            .name(self.to_string())
+            .spawn(move || {
+                let pretty_name = pretty_name_for_thread(&thread::current());
+
+                // Receive a `writer` handle used to pass handles for temporary channels.
+                server
+                    .init_node_writer()
+                    .expect("Couldn't initialialize node writer");
+
+                // Initialize a function that creates a separate instance on the `server` for each
+                // HTTP request.
+                //
+                // TODO(#813): Remove multiple `clone` calls by either introducing `Arc<Mutex<T>>`
+                // or not using Hyper (move to more simple single-threaded server).
+                let generator_server = server.clone();
+                let service = hyper::service::make_service_fn(move |_| {
+                    let connection_server = generator_server.clone();
+                    async move {
+                        Ok::<_, hyper::Error>(hyper::service::service_fn(move |req| {
+                            let request_server = connection_server.clone();
+                            async move { request_server.serve(req).await }
+                        }))
+                    }
+                });
+
+                // Start the HTTP server.
+                let result = block_on(hyper::Server::bind(&server.address).serve(service));
```

Comment:
I think this will block as long as the server is listening on the port, which means that `start` will never return.

Also, I think this should not be using `block_on`, which uses a threadpool executor, but should rather be using Tokio executors as Hyper is based on Tokio. (https://docs.rs/tokio/0.2.16/tokio/runtime/struct.Runtime.html)

Reply:
This is happening inside a newly created thread: https://github.com/project-oak/oak/pull/772/files/327276097e2c3593e513d92efa4f8ce713ab6a0a#diff-475373f331232762d978c36eb8027376R288-R290
Same is happening in Logger node:
https://github.com/project-oak/oak/blob/006b9210e02e6483d298d24ec2940cff394906ee/oak/server/rust/oak_runtime/src/node/logger.rs#L61-L65
Where `logger` is an infinite loop.

Reply:
Done.
Added a `tokio` runtime for executing a Future.
https://github.com/project-oak/oak/pull/772/files#diff-475373f331232762d978c36eb8027376R313-R317